torch/utils/data/dataloader.py

```diff
@@ -55,132 +89,138 @@ class DataLoader(object):
-                is declared as a top level def, outside of that ``__main__`` check. This ensures
-                they are available in workers as well
-                (this is needed since functions are pickled as references only, not bytecode).
-
-              By default, each worker will have its PyTorch seed set to
-              ``base_seed + worker_id``, where ``base_seed`` is a long generated
-              by main process using its RNG. However, seeds for other libraies
-              may be duplicated upon initializing workers (w.g., NumPy), causing
-              each worker to return identical random numbers. (See
-              :ref:`dataloader-workers-random-seed` section in FAQ.) You may
-              use :func:`torch.initial_seed()` to access the PyTorch seed for
-              each worker in :attr:`worker_init_fn`, and use it to set other
-              seeds before data loading.
-
-    .. warning:: If ``spawn`` start method is used, :attr:`worker_init_fn` cannot be an
-                 unpicklable object, e.g., a lambda function.
-
-    The default memory pinning logic only recognizes Tensors and maps and iterables
-    containg Tensors.  By default, if the pinning logic sees a batch that is a custom type
-    (which will occur if you have a ``collate_fn`` that returns a custom batch type),
-    or if each element of your batch is a custom type, the pinning logic will not
-    recognize them, and it will return that batch (or those elements)
-    without pinning the memory.  To enable memory pinning for custom batch or data types,
-    define a ``pin_memory`` method on your custom type(s).
-
-    Example::
-
-        class SimpleCustomBatch:
-            def __init__(self, data):
-                transposed_data = list(zip(*data))
-                self.inp = torch.stack(transposed_data[0], 0)
-                self.tgt = torch.stack(transposed_data[1], 0)
-
-            def pin_memory(self):
-                self.inp = self.inp.pin_memory()
-                self.tgt = self.tgt.pin_memory()
-                return self
-
-        def collate_wrapper(batch):
-            return SimpleCustomBatch(batch)
-
-        inps = torch.arange(10 * 5, dtype=torch.float32).view(10, 5)
-        tgts = torch.arange(10 * 5, dtype=torch.float32).view(10, 5)
-        dataset = TensorDataset(inps, tgts)
-
-        loader = DataLoader(dataset, batch_size=2, collate_fn=collate_wrapper,
-                            pin_memory=True)
-
-        for batch_ndx, sample in enumerate(loader):
-            print(sample.inp.is_pinned())
-            print(sample.tgt.is_pinned())
 
+    .. warning:: If the ``spawn`` start method is used, :attr:`worker_init_fn`
+                 cannot be an unpicklable object, e.g., a lambda function. See
+                 :ref:`multiprocessing-best-practices` on more details related
+                 to multiprocessing in PyTorch.
+
+    .. note:: ``len(dataloader)`` is determined by length of the sampler used.
```

Comment:
This is definitely too low-level. The users should not care about the particular implementation we use. We should say something along the lines of "`__len__` is a heuristic based on the length of the sampler. Note that subclasses of `IterableDataset` always use an infinite sampler, so one should not query this method unless they work with a map-style dataset." This way we can avoid all the problems related to multiprocess sampling from iterable datasets and make it easy to understand.

We should also make it an error to use `__len__` with a non-map dataset.

Reply:
fixed!